Documentation/templates/ib/segues-swift3.md

```diff
@@ -48,10 +49,10 @@ vc.perform(segue: StoryboardSegue.Message.embed)
 
 // or match them (in prepareForSegue):
 override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
-  switch StoryboardSegue.Message(rawValue: segue.identifier!)! {
-  case .embed:
+  switch StoryboardSegue.Message(segue) {
+  case .embed?:
     // Prepare for your custom segue transition
-  case .nonCustom:
+  case .nonCustom?:
     // Pass in information to the destination View Controller
   }
```

Comment:
That won't compile without adding the last missing `case nil` ("switch must be exhaustive" error)

Reply:
Maybe better a `default` case, because our current segues implementation has all the segues in a storyboard, not just the ones starting from 1 segue.

Reply:
yeah, but if that method is called with a segue from another scene, then `StoryboardSegue.Message(segueFromAnotherScene)` will also return `nil`, not being able to construct a `StoryboardSegue.Message` instance with `rawValue = segueFromAnotherScene.identifier`, so… `nil` should work here too

tbh I prefer avoiding `default` when possible, so that if later on the user adds another `segue` from that `Message` scene, the `switch` will warn them that they forgot a `.some(StoryboardScene.Message)` case in there, while a `default` would absorb both segue from another scene but also a missing segue starting from `Message` and forgotten in the `switch`.

Reply:
I think you're mixing up PRs? This is the PR to split scenes and segues templates, not the scene-specific segues.

This PR's is our old's way of generating segue enums, which is just one big enum for all the segues in a storyboard. Our current implementation can not help a user if they try to use a case for an enum that can't happen for this scene, or init an enum from a storyboard where some segues happen to have the same name.

With our current segues, the only thing a user can do is use `default`, because there will be a bunch of segues that won't matter for this scene, as they can never happen.

Reply:
Mmmh maybe I am (tired), but I still don't think `default` would be the right choice?

`default` in your case will both cover `.some(_)` — meaning the `segue` passed will be a valid and existing `StoryboardSegue.Message`, so one of all the segues staring from that `Message` scene but that you have forgotten to cover in your `switch`, right? — and `.none` — meaning a `segue` coming from another scene, or unnamed.

All named segues starting from the scene named `Message` should be handled in the `prepare(for segue: UIStoryboardSegue)` method and thus covered as a dedicated `case` in the `switch` in my opinion in a classic app, right? And if you have a `case` in your `switch StoryboardScene.Message(segue)` for all `cases` listed in the `StoryboardScene.Message` enum, then the only remaining case — that will indeed make the `switch` exhaustive then — would be `case nil`, which will happen either if `segue.identifier` is `nil`, or if the `segue` isn't actually one of the cases of the `StoryboardScene.Message` enum.

So my point is: 

1. Using `case nil:`  here is enough to make the `switch` exhaustive, as both a segue from another scene (and thus an `identifier` not listed in the `cases` of the `enum`)  AND an unnamed segue will make `StoryboardScene.Message(segue)` return `nil`
2. If later on the user adds a named segue starting from the `Message` scene, then `switch StoryboardScene.Message(segue)` will have to explicitly add that `case .theNewSegueIdentifier` as `case nil` won't cover it — and that's good, as if enforces not to forget it — while `case default:` wouldn't enforce anything and let the user forget about handling the case for that new segue in their `switch`.

Reply:
I completely agree that, when possible, exhaustive enums are the way to go. But:
> a segue from another scene (and thus an identifier not listed in the cases of the enum) ... return nil

Again, this is in the other PR, #364. Our current segue enums, **don't** have this information. Our current segue information is for a **whole** storyboard. My frustration with the current situation is what led to me writing #364.

With our current segue info, let's say you have a scene A with 5 segues. But in that storyboard, there are another 15 segues unrelated to that scene (about average case, not too unthinkable). You can't expect users to, for every scene in that storyboard, write a completely exhaustive `switch` statement. Most users just write switches with the cases for that particular scene, and ignore the rest --> `default`.

Does this prevent errors if the user adds more segues to a scene? No. But exhaustively handling every case of a whole storyboard in every scene isn't better. I'd argue it's worse.